### PR TITLE
EVG-7255 add first failing task in version trigger

### PR DESCRIPTION
**Pull request by ablack12**
The Goal: Some users want a notification when the first task fails, not when the whole patch has finished. 

There does seem to be a race condition where tasks that fail around the same time all send the notification (which I think is why I thought this was broken). I'm thinking `generateWithAlertRecord` should do it before actually inserting the alert record because potentially generating takes some time.

### CODE REVIEW

**Comment by ablack12** [comment]
I manually looked at the query for `ByFirstFailureInVersion` for my test versions and I see multiple alerts, where only one should've been created since we shouldn't have created any if one already exists, so I wonder if something else in my understanding of triggers is off.

**Comment by john-m-liu** [comment]
The problem is probably the alertrecord isn't getting created with the right data or we're querying for it wrong. Haven't looked at this in depth yet but we should definitely try to fix the existing trigger and utilize that

**Comment by ablack12** [comment]
This trigger calls taskFirstFailureInVersion, which does seem to do the right thing: Find queries on Type being `FirstVersionFailureId` and VersionId and subscriptionId, and if this doesn't return anything it inserts a new alert record with type `FirstVersionFailureId`, the versionId, and the subscriptionId.  

I don't see anywhere else this type gets called. I'll stash my code and play around a little bit more with the existing trigger that seems wonky. When do we evaluate triggers? It seems interesting that the existing one _does_ only consider the first failing task, but also doesn't do it until the version is done.

File: public/static/js/notifications.js [hunk 1]
```diff
@@ -17,17 +17,21 @@ mciModule.controller('NotificationsController', function($scope, $window, mciUse
...
       }
       $scope.subscriptions = _.filter(resp.data, function(subscription){
         if (
-          subscription.id === patchFinishId || 
+          subscription.id === patchFinishId ||
+          subscription.id === patchFirstFailureId ||
           subscription.id === buildBreakId || 
-          subscription.id === spawnhostExpirationId || 
+          subscription.id === spawnhostExpirationId ||
+          subscription.id === spawnhostOutcomeId ||
```
Comment (john-m-liu) [inline public/static/js/notifications.js]: Thanks for fixing this, I think I told myself to open a ticket about this a long time ago but forgot

File: model/event/subscriptions.go [hunk 2]
```diff
@@ -451,6 +453,8 @@ func CreateOrUpdateImplicitSubscription(resourceType string, id string,
 			switch resourceType {
 			case ImplicitSubscriptionPatchOutcome:
 				temp = NewPatchOutcomeSubscriptionByOwner(user, subscriber)
+			case ImplicitSubscriptionPatchFirstFailure:
+				temp = NewFirstTaskFailureInVersionSubscriptionByOwner(user, subscriber)
```
Comment (john-m-liu) [inline model/event/subscriptions.go]: Did you double check that this doesn't notify for commits where the user is the author? It looks like the task triggers will match on owner if the task's version was authored by that person

Comment (ablack12) [reply model/event/subscriptions.go]: By commits do you mean mainline builds vs patches? Will selecting on PatchRequester work with that? (I understand vaguely how the Subscription works but if you have a good description that would also be good)

Comment (john-m-liu) [reply model/event/subscriptions.go]: Yep exactly. How it works is actually pretty elegant - https://github.com/evergreen-ci/evergreen/blob/master/model/event/subscriptions.go#L139 will run the triggers for a given file in the triggers package if an event is logged and the user has selectors that are a subset of the selectors for the trigger. So I think what you changed should work

**Pull request merged**